cmd/restic/cmd_stats.go

```diff
@@ -0,0 +1,296 @@
+			stats.uniqueFiles[fid] = struct{}{}
+
+			if countMode == countModeUniqueFilesByContent {
+				// simply count the size of each unique file (unique by contents only)
+				stats.TotalSize += node.Size
+				stats.TotalFileCount++
+			}
+			if countMode == countModeBlobsPerFile {
+				// count the size of each unique blob reference, which is
+				// by unique file (unique by contents and file path)
+				for _, blobID := range node.Content {
+					// ensure we have this file (by path) in our map; in this
+					// mode, a file is unique by both contents and path
+					nodePath := filepath.Join(npath, node.Name)
+					if _, ok := stats.fileBlobs[nodePath]; !ok {
+						stats.fileBlobs[nodePath] = restic.NewIDSet()
+						stats.TotalFileCount++
+					}
+					if _, ok := stats.fileBlobs[nodePath][blobID]; !ok {
+						// is always a data blob since we're accessing it via a file's Content array
+						blobSize, found := repo.LookupBlobSize(blobID, restic.DataBlob)
+						if !found {
+							return true, fmt.Errorf("blob %s not found for tree %s", blobID, *node.Subtree)
+						}
+
+						// count the blob's size, then add this blob by this
+						// file (path) so we don't double-count it
+						stats.TotalSize += uint64(blobSize)
+						stats.fileBlobs[nodePath].Insert(blobID)
+
+						// this mode also counts total unique blob _references_ per file
+						stats.TotalBlobCount++
+					}
+				}
+			}
+		}
+	}
+
+	if countMode == countModeRestoreSize {
+		// as this is a file in the snapshot, we can simply count its
+		// size without worrying about uniqueness, since duplicate files
+		// will still be restored
+		stats.TotalSize += node.Size
+		stats.TotalFileCount++
+	}
+
+	return true, nil
+}
+
+// makeFileIDByContents returns a hash of the blob IDs of the
+// node's Content in sequence.
+func makeFileIDByContents(node *restic.Node) fileID {
+	var bb []byte
+	for _, c := range node.Content {
+		bb = append(bb, []byte(c[:])...)
+	}
+	return sha256.Sum256(bb)
+}
+
+func verifyStatsInput(gopts GlobalOptions, args []string) error {
+	// require a recognized counting mode
+	switch countMode {
+	case countModeRestoreSize:
+	case countModeUniqueFilesByContent:
+	case countModeBlobsPerFile:
+	case countModeRawData:
+	default:
+		return fmt.Errorf("unknown counting mode: %s (use the -h flag to get a list of supported modes)", countMode)
+	}
+
+	// ensure at most one snapshot was specified
+	if len(args) > 1 {
+		return fmt.Errorf("only one snapshot may be specified")
+	}
+
+	// if a snapshot was specified, mark it as the one to scan
+	if len(args) == 1 {
+		snapshotIDString = args[0]
+	}
+
+	return nil
+}
+
+// statsContainer holds information during a walk of a repository
+// to collect information about it, as well as state needed
+// for a successful and efficient walk.
+type statsContainer struct {
+	TotalSize      uint64 `json:"total_size"`
+	TotalFileCount uint64 `json:"total_file_count"`
+	TotalBlobCount uint64 `json:"total_blob_count,omitempty"`
+
```

Comment:
We no longer have an "idSet" field here because I rely on the Walk function to add each node to the set of ignored trees so we only visit each one once.